### ISSUE TITLE
Heterogeneous Convolutional Wrapper example from documentation returns an error

### ISSUE THREAD

**Issue by igormorgado** [pyg-team/pytorch_geometric#8121]
### 🐛 Describe the bug

Following the example presented at [Heterogeneous Convolution Wrapper](https://pytorch-geometric.readthedocs.io/en/latest/tutorial/heterogeneous.html#using-the-heterogeneous-convolution-wrapper), reproduced below

```python
import torch_geometric.transforms as T
from torch_geometric.datasets import OGB_MAG
from torch_geometric.nn import HeteroConv, GCNConv, SAGEConv, GATConv, Linear


dataset = OGB_MAG(root='./data', preprocess='metapath2vec', transform=T.ToUndirected())
data = dataset[0]

class HeteroGNN(torch.nn.Module):
    def __init__(self, hidden_channels, out_channels, num_layers):
        super().__init__()

        self.convs = torch.nn.ModuleList()
        for _ in range(num_layers):
            conv = HeteroConv({
                ('paper', 'cites', 'paper'): GCNConv(-1, hidden_channels),
                ('author', 'writes', 'paper'): SAGEConv((-1, -1), hidden_channels),
                ('paper', 'rev_writes', 'author'): GATConv((-1, -1), hidden_channels),
            }, aggr='sum')
            self.convs.append(conv)

        self.lin = Linear(hidden_channels, out_channels)

    def forward(self, x_dict, edge_index_dict):
        for conv in self.convs:
            x_dict = conv(x_dict, edge_index_dict)
            x_dict = {key: x.relu() for key, x in x_dict.items()}
        return self.lin(x_dict['author'])

model = HeteroGNN(hidden_channels=64, out_channels=dataset.num_classes,
                  num_layers=2)
```

The model returns a `ValueError` at the convolution method defined to the edge `paper, rev_writes, author`. The error states that setting `True`  to  the `add_self_loops`   (the default), will lead to incorrect MP results.

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[19], line 1
----> 1 model = HeteroGNN(hidden_channels=64, out_channels=dataset.num_classes, num_layers=2)

Cell In[18], line 7, in HeteroGNN.__init__(self, hidden_channels, out_channels, num_layers)
      5 self.convs = torch.nn.ModuleList()
      6 for _ in range(num_layers):
----> 7     conv = HeteroConv({
      8         ('paper', 'cites', 'paper'): GCNConv(-1, hidden_channels),
      9         ('author', 'writes', 'paper'): SAGEConv((-1, -1), hidden_channels),
     10         ('paper', 'rev_writes', 'author'): GATConv((-1, -1), hidden_channels),
     11     }, aggr='sum')
     12     self.convs.append(conv)
     14 self.lin = Linear(hidden_channels, out_channels)

File venv/lib/python3.11/site-packages/torch_geometric/nn/conv/hetero_conv.py:57, in HeteroConv.__init__(self, convs, aggr)
     54 super().__init__()
     56 for edge_type, module in convs.items():
---> 57     check_add_self_loops(module, [edge_type])
     59 src_node_types = set([key[0] for key in convs.keys()])
     60 dst_node_types = set([key[-1] for key in convs.keys()])

File venv/lib/python3.11/site-packages/torch_geometric/utils/hetero.py:54, in check_add_self_loops(module, edge_types)
     52 is_bipartite = any([key[0] != key[-1] for key in edge_types])
     53 if is_bipartite and getattr(module, 'add_self_loops', False):
---> 54     raise ValueError(
     55         f"'add_self_loops' attribute set to 'True' on module '{module}' "
     56         f"for use with edge type(s) '{edge_types}'. This will lead to "
     57         f"incorrect message passing results.")

ValueError: 'add_self_loops' attribute set to 'True' on module 'GATConv((-1, -1), 64, heads=1)' for use with edge type(s) '[('paper', 'rev_writes', 'author')]'. This will lead to incorrect message passing results.
```

AFAIK, to also consider the node itself in the update function in  GAT architecture,  the self loops are needed. Hence my question about the error.

IMHO, the documentation should be updated to reflect a workable example.  And my questions here are:

-  why this error is being raised?
-  Why self loops here will lead to incorrect message passing?
-  What should be the correct approach?

Best regards

### Environment

* PyG version: 2.3.1
* PyTorch version: 2.0.1+cu117
* OS: Arch Linux
* Python version: 3.11.5
* CUDA/cuDNN version: CUDA 12.2 / cuDNN: 8.9.2.26-1
* How you installed PyTorch and PyG (`conda`, `pip`, source): pip (in venv)
* Any other relevant information (*e.g.*, version of `torch-scatter`): I don t think so

**Comment by rusty1s** [comment]
Thanks for spotting this. This is indeed a mistake in our documentation. I made this more clear in #8147.

TL;DR: We cannot add self-loops in bipartite graphs (edge types where source and destination node types are different), and so `GATConv` would mistakenly add `[(0, 0), (1, 1), ...]` edges to the bipartite graph (which is not correct). The correct approach is to use `add_self_loops=False`. If you want to preserve central node information in `HeteroConv`, you would currently need to wrap `GATConv`:
```python
class MyGATConv(GATConv):
     def __init__(self, *args, **kwargs):
          super().__init__(*args, **kwargs)
          self.root_lin = Linear(self.in_channels[1], self.out_channels)
          
      def forward(self, x: Tuple[Tensor, Tensor], edge_index: Tensor):
           return super().forward(x, edge_index) + self.root_lin(x[1])
```
and use `MyGATConv` instead of `GATConv` in `HeteroConv`.
